src/main.cpp

```diff
@@ -5205,6 +5234,159 @@ bool static ProcessMessage(CNode* pfrom, string strCommand, CDataStream& vRecv,
+                    if (!partialBlock.IsTxAvailable(i))
+                        req.indexes.push_back(i);
+                }
+                if (req.indexes.empty()) {
+                    vector<CTransaction> empty_txn;
+                    CBlock block;
+                    status = partialBlock.FillBlock(block, empty_txn);
+                    if (status == READ_STATUS_OK) {
+                        CValidationState state;
+                        ProcessNewBlock(state, chainparams, pfrom, &block, false, NULL);
+                        int nDoS;
+                        if (state.IsInvalid(nDoS)) {
+                            assert (state.GetRejectCode() < REJECT_INTERNAL); // Blocks are never rejected with internal reject codes
+                            pfrom->PushMessage(NetMsgType::REJECT, strCommand, (unsigned char)state.GetRejectCode(),
+                                               state.GetRejectReason().substr(0, MAX_REJECT_MESSAGE_LENGTH), block.GetHash());
+                            if (nDoS > 0) {
+                                LOCK(cs_main);
+                                Misbehaving(pfrom->GetId(), nDoS);
+                            }
+                        }
+                    } else {
+                        Misbehaving(pfrom->GetId(), 100); // There should be no way to hit this
+                        LogPrintf("Peer %d sent us invalid compact block which we should have been able to reconstruct", pfrom->id);
+                    }
+                } else {
+                    req.blockhash = pindex->GetBlockHash();
+                    pfrom->PushMessage(NetMsgType::GETBLOCKTXN, req);
+                }
+            }
+        } else {
+            // Dirty hack to process as if it were just a headers message
+            std::vector<CBlock> headers;
+            headers.push_back(cmpctblock.header);
+            CDataStream vHeadersMsg(SER_NETWORK, PROTOCOL_VERSION);
+            vHeadersMsg << headers;
+            return ProcessMessage(pfrom, NetMsgType::HEADERS, vHeadersMsg, nTimeReceived, chainparams);
+        }
+
+        CheckBlockIndex(chainparams.GetConsensus());
+    }
+
+    else if (strCommand == NetMsgType::BLOCKTXN && !fImporting && !fReindex) // Ignore blocks received while importing
+    {
+        BlockTransactions resp;
+        vRecv >> resp;
+
+        map<uint256, pair<NodeId, list<QueuedBlock>::iterator> >::iterator it = mapBlocksInFlight.find(resp.blockhash);
+        if (it == mapBlocksInFlight.end() || !it->second.second->partialBlock ||
+                it->second.first != pfrom->GetId()) {
+            LogPrint("net", "Peer %d sent us block transactions for block we didn't requset", pfrom->id); //...or we timed out
+            return true;
+        }
+
+        PartiallyDownloadedBlock& partialBlock = *it->second.second->partialBlock;
+        CBlock block;
+        ReadStatus status = partialBlock.FillBlock(block, resp.txn);
```

Comment:
For now I swapped in the same call-headers-msg code from CMPCTBLOCK handling into the CMPCTBLOCK code  call ::BLOCKTXN handling. Its a dirty hack to serialize and then immediately deserialize a message, but its a short message so it doesnt cost much, and I'm assuming the network stack will be (properly) rewritten soonish anyway, so then it can be called like the individual function it should be.